### DISCUSSION TITLE
How to export generated keypairs?

### DISCUSSION THREAD

**Discussion by MrWildanMD** [Mastercard/pkcs11-tools#57] [Q&A] [answered]
I'm using Utimaco HSM Simulator but how i can export the generated keypairs?
i need to export the private and the public generated by hsm to use it later with CA

**Comment by keldonin** [comment]
Hi @MrWildanMD,

Can you please elaborate more on your plan? do you mean to deploy your keys into an HSM?
On a side note, the simulator does not have the same level of assurance than a real HSM.

**Reply by MrWildanMD** [reply]
Hello @keldonin , sorry for late reply, so here the flow
So the keypair generation is done inside the HSM, then after it generated i need to export the keypair and the CSR to use it later in CA services.
Because currently i have CA service that doing the keypair generation and CSR, that i need it to be generated inside HSM not in the CA services itself. Thanks

**Reply by MrWildanMD** [reply]
And of course the simulator i'm using only for the testing purpose before doing it in the physical hsm devices 😃

**Reply by keldonin** [reply]
Thanks for the comments.

It's effectively unusual to have CA service that also generates the key (especially hardware keys).

Anyway, if you want to export a key, you will need to:
 - at key creation, ensure that the key has `extractable` attribute set to `true`. If you miss that part, you won't be able to extract your key. Using the toolkit, this can be achieved this way (I am omitting connectivity arguments):
   ```
   $ p11keygen -k rsa -b 2048 -i my-key sign verify extractable
   ```
   (*note: `sign`and `verify` are set as you mentioned you need to generate a CSR with this key*)
   
 - later, on your target HSM, you will need to generate a wrapping key:
   ```
   $ p11keygen -k rsa -b 4096 -i wrapping-key wrap unwrap
   $ p11cat pubk/wrapping-key > wrapping-key.pubk
   ```
 - back to the source HSM, you need to import that public key:
   ```
   $ p11importpubk -f wrapping-key.pubk -i wrapping-key wrap
   ```
     
 - at this point, you are ready to perform wrapping, on the source HSM:
   ```
   $ p11wrap -w wrapping-key -i my-key -a envelope > my-key-wrapped-under-wrapping-key.txt
   ```
   
 - import (unwrap) the exported key on your target HSM:
   ```
   $ p11unwrap -f my-key-wrapped-under-wrapping-key.txt
   ```
 
You can check out documentation for more details: [link](https://github.com/Mastercard/pkcs11-tools/blob/master/docs/MANUAL.md#exchanging-a-keys-between-tokens---the-long-way)

There is also a way to do this faster, as `p11keygen` can combine both the key generation and wrapping under a public key. See [this link](https://github.com/Mastercard/pkcs11-tools/blob/master/docs/MANUAL.md#exchanging-keys-between-tokens---the-accelerated-way) for more information.

Finally, remember that extractable keys are vulnerable: if anyone can have access to your HSM, he/she will be able to wrap it under a key of his/her choice and eventually recover/compromize the key. You should limit the usage of extractable keys as much as possible, and ensure you have compensating controls in your environment.

**Reply by MrWildanMD** [reply]
Okay thanks for the explanation, but i want to make sure the step you provided is only for public key right? can i wrap the private key also to export it? so it will be public and private key at same time exported, i have set exportable to true but when i try to cat the private it says private objects cannot be disclosed, thanks

**Reply by MrWildanMD** [reply]
Because i need the private key for the signing purpose (e.g: signing document)

**Reply by keldonin** [reply]
The explanation provided is for the private key. Note that the public key if present will also be carried inside the wrap file and imported back onto the target HSM.

You can't display the content of an extractable key directly. This attribute means the key can be wrapped.

**Reply by MrWildanMD** [reply]
Okay, thanks for the explanation, can i have a feature request? like signing data, verify signed data, or if i can contribute to this can you explain the project dir? so i can contribute 😄

**Reply by keldonin** [reply]
Hi @MrWildanMD , sorry for the late reply. You can request features, obviously. that being said, `pkcs11-tools` is meant to be a key management tool, in the first place. Signing or verifying data is shifting the intent towards key operation, not sure if it belongs here.

Can you be more specific on the use case? what data would you like to sign/verify? In what context?

Cheers,

**Reply by MrWildanMD** [reply]
sorry for late reply, so i would like to use the private key to sign into docs (like pdf) so the pdf can be verfied using the p12.
if u confused what i'm saying, the content of private key will be injected inside docs then within p12 from keygen u can verify it

**Reply by keldonin** [reply]
I must admit this is confusing indeed. A PKCS\#12 usually contains a private key together with certificates. To validate documents, you only need a public key (embedded into a certificate + its chain). It means you don't need an HSM or even a PKCS\#12 keystore to do that. Am I missing something?

**Comment by MrWildanMD** [comment]
@keldonin hi, what about luna hsm? when i try ./with_luna p11slotinfo it return token not present, did i miss some step?

**Reply by keldonin** [reply]
`pkcs11-tools` works well with Luna HSM. 
- you need of course to have a working environment. Make sure you can successfully use `ckdemo` from the lunaclient environment. Please consult vendor documentation if that doesn't work. Does your user belong to the `hsmusers` group?
- `with_luna` infers some parameters from environment variables. Please check, these are inside the headers. If you don't provide values, defaults are taken. There is an easy way to show what variables are set:
```bash
$ ./with_luna env | grep PKCS11
```
You can adjust these variables by writting them into a `.pkcs11rc` file, sitting in your home directory.

Hope this helps,

**Reply by MrWildanMD** [reply]
<img width="772" alt="Screenshot 2024-04-26 at 15 11 23" src="https://github.com/Mastercard/pkcs11-tools/assets/60354284/6dfa8c34-d6dd-42f3-a2b3-5239b7675c9e">
@keldonin As you can see my `ckdemo` is running well but when i try to `p11slotinfo` it return error

**Reply by keldonin** [reply]
can you please choose "11" (Slot Info) and report what you get with `ckdemo` ?

**Reply by MrWildanMD** [reply]
@keldonin here you go
<img width="1440" alt="image" src="https://github.com/Mastercard/pkcs11-tools/assets/60354284/08b63aa1-d1ac-4e2c-b48f-326dd782588a">

**Reply by keldonin** [reply]
Alright, can you now execute "Token Info" (12) and select slot `0` when asked for a slot number, and report the content?

As an example, here is what I have on my config:
```
$ ckdemo -n
ckdemo (64-bit) v10.7.0-255. Copyright (c) 2023 Thales Group. All rights reserved.

ckdemo is the property of Thales Group and is provided to our customers for
diagnostic and development purposes only.  It is not intended for use in
production installations.  Any re-distribution of this program in whole or
in part is a violation of the license agreement.

Modified on Dec 18 2023 at 15:32:15

Starting CHRYSTOKI DEMO - SIMULATION LAB

Status: Doing great, no errors (CKR_OK)

(TITLE) menu titles, (99 or FULL) Full Help, (NONE) No help, (0 or EXIT) Quit


Status: Doing great, no errors (CKR_OK)
Enter your choice : 12

Slots available:
	slot#0 - Net Token Slot
	slot#1 - Net Token Slot
	slot#2 - Net Token Slot
Select a slot: 0

Token Info:
	Token label -> lun0woo2.caectf.com-d99-p000
	Token Manufacturer -> Safenet, Inc.
	Token Model -> LunaSA 7.7.1
	Token Serial Number -> XXXXXXXXXXXX
	Token Flags ->
		CKF_RNG
		CKF_LOGIN_REQUIRED
		CKF_USER_PIN_INITIALIZED
		CKF_RESTORE_KEY_NOT_NEEDED
		CKF_TOKEN_INITIALIZED
	Sessions (count/max) -> 0 / 0
	Read/Write Sessions (count/max) -> 0 / 0
	PIN Length (min - max) -> 8 - 255
	Public Memory (free / total) -> 5564691 - 5585635
	Private Memory (free / total) -> 5564691 - 5585635
	Hardware Version                -> 0.0
	Firmware Version (CK_VERSION)   -> 7.7
	Firmware Version (Luna Version) -> 7.7.1
	Bootloader Version              -> 1.1.2
	UTC Time -> 20240502091818
Status: Doing great, no errors (CKR_OK)

(TITLE) menu titles, (99 or FULL) Full Help, (NONE) No help, (0 or EXIT) Quit


Status: Doing great, no errors (CKR_OK)
Enter your choice : 0

Ending CHRYSTOKI DEMO - SIMULATION LAB
```

Can you also provide the output of lunacm ?

```
$ lunacm -q slot list
```

**Reply by MrWildanMD** [reply, marked as answer]
@keldonin the issue is Crystoki.conf permission because i logged in as normal user so i changed the permission to 755 and can run smoothly right now

**Reply by keldonin** [reply]
Thanks for your feedback @MrWildanMD , glad you could find the root cause to your problem.
